composer/checkpoint/save.py

```diff
@@ -0,0 +1,133 @@
+        else:
+            path_saved = None
+
+    return path_saved
+
+
+def _save_sharded_state_dict_to_disk(
+    state_dict: Dict[str, Any],
+    destination_file_path: str,
+    overwrite: bool = False,
+    save_format: str = 'pt',  # or safetensor
+) -> Optional[str]:
+
+    if save_format != 'pt':
+        raise NotImplementedError(f'Saving sharded state dict to disk in format {save_format} is not supported.')
+
+    if state_dict == {}:
+        return None
+
+    # If used specifies filename instead of directory suffixes, strip them and warn.
+    if len(Path(destination_file_path).suffixes) > 0:
+        stripped_path = _strip_suffixes(destination_file_path)
+        warnings.warn(
+            textwrap.dedent(
+                f"""Sharded checkpoints require a directory path not a file path:
+            {destination_file_path} will have its extensions stripped and checkpoints will be saved in {stripped_path}
+            as {stripped_path}/{_TORCH_DISTRIBUTED_CHECKPOINTS_FILENAME}""",
+            ),
+        )
+        destination_file_path = stripped_path
+
+    if dist.get_global_rank() == 0 and not overwrite and os.path.exists(destination_file_path):
+        raise ValueError(f'Directory {destination_file_path} already exists. Set overwrite=True to overwrite it.')
+
+    log.debug(
+        f'Starting saving of sharded state dict to {destination_file_path}/{_TORCH_DISTRIBUTED_CHECKPOINTS_FILENAME}',
+    )
+
+    # For 2.3.0 and above you can use checkpoint_id, but this version works the best for all versions
+    # of torch (and makes pyright happier) that we support, so we use it for now.
+    DCP.save(state_dict=state_dict, storage_writer=DCP.FileSystemWriter(destination_file_path))
+
+    return destination_file_path + '/' + _TORCH_DISTRIBUTED_CHECKPOINTS_FILENAME
+
+
+def _save_full_state_dict_to_disk(
+    state_dict: Dict[str, Any],
+    destination_file_path: str,
+    overwrite: bool = False,
+    save_format: str = 'pt',  # or hf, safetensor
+) -> Optional[str]:
+
+    if save_format != 'pt':
+        raise NotImplementedError(f'Saving full state dict to disk in format {save_format} is not supported.')
+
+    if not overwrite and os.path.exists(destination_file_path):
+        raise ValueError(f'File {destination_file_path} already exists. Set overwrite=True to overwrite it.')
+
+    if dist.get_global_rank() == 0:
+        _write_checkpoint_file(state_dict=state_dict, filename=destination_file_path)
+        return destination_file_path
+    return None
+
+
+def is_state_dict_sharded(state_dict: Dict[str, Any]) -> bool:
+    """Determines if the state dict is sharded.
+
+    Args:
+        state_dict (Dict[str, Any]): The state dict to check.
+
+    Returns:
+        bool: Whether the state dict is sharded.
+    """
+    sample_value = next(iter(state_dict.values()))
+    return isinstance(sample_value, ShardedTensor) or isinstance(sample_value, DTensor)
```

Comment:
i think even for sharded state dict, it's possible there is Tensor in state_dict ? So should be better to just pass in `is_sharded` to  `save_state_dict_to_disk`

Reply:
I don't think that is the case

Reply:
I found this case while working on this issue: https://github.com/pytorch/pytorch/issues/125740. That issues is caused by deduping `Tensor` in state_dict to different ranks. Specifically, it's a optimizer scalar tensor in optimizer state

Reply:
oh I see yeah ok well we can test all values then maybe? and we can add a unit test with optimizers too?

Reply:
> we can test all values

I think that works. but why not just pass in the `is_sharded` ?  that should be much easier and safe ? :)

Reply:
would be nice to not have to pass a flag if one could infer it